include/tvm/relay/type.h

```diff
@@ -149,6 +150,63 @@ class TypeVarNode : public TypeNode {
 
 RELAY_DEFINE_NODE_REF(TypeVar, TypeVarNode, Type);
 
+/*!
+ * \brief A global type variable that is used for defining new types or type aliases.
+ */
+class GlobalTypeVar;
+/*! \brief GlobalTypeVar container node */
+class GlobalTypeVarNode : public TypeNode {
+ public:
+  /*!
+   * \brief The variable itself is only meaningful when
+   *  kind is ShapeVar; otherwise, we only use the name.
+   */
+  tvm::Var var;
+  /*! \brief The kind of type parameter */
+  Kind kind;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+    v->Visit("var", &var);
+    v->Visit("kind", &kind);
+    v->Visit("span", &span);
+  }
+
+  TVM_DLL static GlobalTypeVar make(std::string name, Kind kind);
+
+  static constexpr const char* _type_key = "relay.GlobalTypeVar";
+  TVM_DECLARE_NODE_TYPE_INFO(GlobalTypeVarNode, TypeNode);
+};
+
+RELAY_DEFINE_NODE_REF(GlobalTypeVar, GlobalTypeVarNode, Type);
+
+/*!
+ * \brief Type application.
+ */
+class TypeCall;
+/*! \brief TypeCall container node */
+class TypeCallNode : public TypeNode {
```

Comment:
@MarisaKirisame can you better explain how type calls are supposed to work? I mean both in terms of how they're used to type ADTs (namely, can you outline an example that we may want to use for docs?) and how the Relay representation is supposed to work. The tests for it seem to fail and I'm having trouble understanding it

Reply:
Also now that I think of it, we'll have to add all these new types and expressions to the Relay language ref, so it's good to take the time to write good descriptions.

Reply:
TypeCall(nat, {}) is natural number
TypeCall(List, Int) is list of int
etc.

Reply:
I see, so is it correct to say type calls represent ADTs with their type parameters?

Reply:
Wider question: Should type calls still be called type calls? Their purpose is to list out the type parameters to a polymorphic ADT (see Marisa's list example above) and nothing more. They are called type calls because the type data struct can be considered a type-level lambda (taking the type params as arguments, returning the global type var), but this seems rather oblique in terms of being descriptive names

Reply:
Yeah, renaming it ADTType or something similar sounds good.

Reply:
I'll do the renaming then